### ISSUE TITLE
C++ Include paths not found for framework packages

### ISSUE THREAD

**Issue by graph** [microsoft/vscode-cpptools#2234]
Type: LanguageService
**----- Input information below -----**

**Describe the bug**
- OS and Version: Mac
- VS Code Version: 1.25
- C/C++ Extension Version: 0.17.6
- A clear and concise description of what the bug is.

Can't find includes in frameworks.

**To Reproduce**

1. add include path in c_cpp_properties.json that refers to a SomeLibrary.framework
2. see include errors by the intellisense feature.

frameworks are included like

#include <SomeLibrary/SomeLibrary.h>

**Expected behavior**
include paths to be found for SomeLibrary/SomeLibrary.h

**Additional context**
The structure inside the SomeLibrary.framework directory has only a directory called "headers" where all the headers are listed. Directory structure for frameworks do not match how they are included in C++. When compiling everything is obviously fine. My codebase is littered with these errors.

**Comment by bobbrow** [comment]
I'm not seeing what you're seeing.  Can you provide some additional information or share a repro with us?  Do you have `compilerPath` or `macFrameworksPath` set in `c_cpp_properties.json`?

![image](https://user-images.githubusercontent.com/12818240/42392246-0424e412-8107-11e8-83f1-f2f04c4db18a.png)

**Comment by graph** [comment]
the default frameworks is fine. I'm talking about a proprietory framework that is included in the project. I didn't know about `macFrameworksPath` I will look into that and report back.

**Comment by bobbrow** [comment]
Ah.  If this is your own framework, then add the parent folder of `SomeLibrary.framework` to the `macFrameworksPath` and you should be good to go.

**Comment by graph** [comment]
I tried the following:

```
            "macFrameworkPath": [
                "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks",
                "${workspaceFolder}/MyProject/FrameworksDir"
            ],
```

And it's not working. I'm looking at the [documentation here](https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/c_cpp_properties.json.md)

I'm including the framework header in .cpp files. This include thing is feeling kind of buggy to me. I have also added

```
            "includePath": [
                "${workspaceFolder}/MyProject/**",
                "${workspaceFolder}/Pods/**"
            ],
```

And most pod includes are found but some aren't. I'm new to vs code it could be I'm not using it correctly. For the pods that it says not found I can still CMD+Click them and it jumps to the correct header. I can't CMD+Click on `#include` of private framework though.

**Comment by bobbrow** [comment]
What does the folder layout look like?  There should be a folder or symlink in `${workspaceFolder}/MyProject/FrameworksDir/SomeLibrary.framework` called `Headers` that points to the location of the header files.  Does your framework have it?  If not, the extension is probably skipping that path.

When you say that some includes aren't found, do you mean there are red or green squiggly lines under them?  If you set the `"C_Cpp.loggingLevel"` setting to `"Debug"` the extension will log what it's doing.

CMD+Click is powered by a different bit of code than the squiggly lines, so it is possible for us to still find the files even with red squiggles present in the file.  Red squiggles would mean that we didn't pick the right set of include paths for your file, so it would be good for us to understand why that is the case.  If there is any code you can share with us or even anonymized folder structures, that would help us.  I suspect the Framework you're using is not shareable though.

**Comment by graph** [comment]
I see red squiggles for some of the cocoapod includes that aren't found. and SomeLibrary.framework does have a folder called `Headers` that points to location of headers. I left work. Next week I'll try "Debug" and report back.

**Comment by graph** [comment]
I enabled the debug and restarted. I see lines that say "SomeFramework.framework/A/Headers" will be indexed". In my private framework there is no "A" folder. just a "MyFramework.framework/Headers/*.{h,hpp}". Maybe the plugin doesn't check that? I modified the folder structure in the framework manual to have an "A/Headers" and restarted vs code that didn't help. 

I see all sorts of lines saying "Processing folder (recursive): /....SomeFramework.framework/Versions/A/Headers" and etc.. but none for my private frameworks. Ditto for ".... Will be indexed"

**Comment by bobbrow** [comment]
It doesn't need an `A` folder.  `MyFramework.framework/Headers` should work.  The only difference between the system libraries and yours appears to be that `Headers` is not a symlink in your case.  But our code is not checking whether it's a symlink or not.

I'll add a debug string in that code and see if we can figure out what's going on in the next release.

**Comment by bobbrow** [comment]
Nevermind, I just set up a repro and found the bug.  `${workspaceFolder}` is not being resolved for some reason.  Can you try using the absolute path instead of the variable?  I'll work on a fix for this.

**Comment by graph** [comment]
Full path works in removing the errors. I noticed another couple of issues

- Autocomplete for include in private library doesn't work but I don't get any errors with full path in c_cpp_properties.json
- There are multiple headers in the framework, only the one that matches the framework name gives no errors. I have hpp files in the framework and those give errors if I include them.

**Comment by bobbrow** [comment]
I opened #2251 for the first issue, but I don't repro the second issue.  Can you open a new issue with some more details, maybe?

I created `MyF.framework/Headers` and put `a.h` and `b.hpp` in it then included them like this:
```
#include <MyF/a.h>
#include <MyF/b.hpp>
```
and I don't get any squiggles.  The declarations from the headers are available in the source file so I know it's working.

**Comment by graph** [comment]
The second point I also can't reproduce anymore. I guess it just took a while for the indexer to index the files. Thanks so much! Your support for a free product is amazing. I look forward for next release :)

**Comment by sean-mcmanus** [comment]
The issues with the Mac framework path have been fixed with 0.17.7.
